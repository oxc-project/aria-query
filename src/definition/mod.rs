use phf::OrderedMap;
#[cfg(test)]
use serde::{ser::SerializeStruct, Serialize};

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAAbstractRole {
    Command,
    Composite,
    Input,
    Landmark,
    Range,
    RoleType,
    Section,
    SectionHead,
    Select,
    Structure,
    Widget,
    Window,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAWidgetRole {
    Button,
    CheckBox,
    GridCell,
    Link,
    MenuItem,
    MenuItemCheckBox,
    MenuItemRadio,
    Option,
    ProgressBar,
    Radio,
    ScrollBar,
    SearchBox,
    Slider,
    SpinButton,
    Switch,
    Tab,
    TabPanel,
    TextBox,
    TreeItem,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIACompositeWidgetRole {
    ComboBox,
    Grid,
    ListBox,
    Menu,
    MenuBar,
    RadioGroup,
    TabList,
    Tree,
    TreeGrid,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIADocumentStructureRole {
    Application,
    Article,
    BlockQuote,
    Caption,
    Cell,
    ColumnHeader,
    Definition,
    Deletion,
    Directory,
    Document,
    Emphasis,
    Feed,
    Figure,
    Generic,
    Group,
    Heading,
    Img,
    Insertion,
    List,
    ListItem,
    Mark,
    Math,
    Meter,
    None,
    Note,
    Paragraph,
    Presentation,
    Row,
    RowGroup,
    RowHeader,
    Separator,
    Strong,
    Subscript,
    Superscript,
    Table,
    Term,
    Time,
    ToolBar,
    ToolTip,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIALandmarkRole {
    Banner,
    Complementary,
    ContentInfo,
    Form,
    Main,
    Navigation,
    Region,
    Search,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIALiveRegionRole {
    Alert,
    Log,
    Marquee,
    Status,
    Timer,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAWindowRole {
    AlertDialog,
    Dialog,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAUncategorizedRole {
    Code,
}

// please define enum equal to the above flow type
#[cfg_attr(test, derive(Serialize))]
pub enum ARIADPubRole {
    DocAbstract,
    DocAcknowledgments,
    DocAfterword,
    DocAppendix,
    DocBacklink,
    DocBiblioentry,
    DocBibliography,
    DocBiblioref,
    DocChapter,
    DocColophon,
    DocConclusion,
    DocCover,
    DocCredit,
    DocCredits,
    DocDedication,
    DocEndnote,
    DocEndnotes,
    DocEpigraph,
    DocEpilogue,
    DocErrata,
    DocExample,
    DocFootnote,
    DocForeword,
    DocGlossary,
    DocGlossref,
    DocIndex,
    DocIntroduction,
    DocNoteref,
    DocNotice,
    DocPagebreak,
    DocPagelist,
    DocPart,
    DocPreface,
    DocPrologue,
    DocPullquote,
    DocQna,
    DocSubtitle,
    DocTip,
    DocToc,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAGraphicsRole {
    GraphicsDocument,
    GraphicsObject,
    GraphicsSymbol,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIARole {
    ARIAWidgetRole(ARIAWidgetRole),
    ARIACompositeWidgetRole(ARIACompositeWidgetRole),
    ARIADocumentStructureRole(ARIADocumentStructureRole),
    ARIALandmarkRole(ARIALandmarkRole),
    ARIALiveRegionRole(ARIALiveRegionRole),
    ARIAWindowRole(ARIAWindowRole),
    ARIAUncategorizedRole(ARIAUncategorizedRole),
    ARIADPubRole(ARIADPubRole),
    ARIAGraphicsRole(ARIAGraphicsRole),
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIARoleDefinitionKey {
    ARIAAbstractRole(ARIAAbstractRole),
    ARIARole(ARIARole),
    ARIADPubRole(ARIADPubRole),
    ARIAGraphicsRole(ARIAGraphicsRole),
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIANameFromSources {
    Author,
    Contents,
    Prohibited,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAState {
    AriaBusy,
    AriaChecked,
    AriaDisabled,
    AriaExpanded,
    AriaGrabbed,
    AriaHidden,
    AriaInvalid,
    AriaPressed,
    AriaSelected,
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAProperty {
    AriaActivedescendant,
    AriaAtomic,
    AriaAutocomplete,
    AriaBraillelabel,
    AriaBrailleroledescription,
    AriaColcount,
    AriaColindex,
    AriaColspan,
    AriaControls,
    AriaCurrent,
    AriaDescribedby,
    AriaDescription,
    AriaDetails,
    AriaDropeffect,
    AriaErrormessage,
    AriaFlowto,
    AriaHaspopup,
    AriaKeyshortcuts,
    AriaLabel,
    AriaLabelledby,
    AriaLevel,
    AriaLive,
    AriaModal,
    AriaMultiline,
    AriaMultiselectable,
    AriaOrientation,
    AriaOwns,
    AriaPlaceholder,
    AriaPosinset,
    AriaReadonly,
    AriaRelevant,
    AriaRequired,
    AriaRoledescription,
    AriaRowcount,
    AriaRowindex,
    AriaRowspan,
    AriaSetsize,
    AriaSort,
    AriaValuemax,
    AriaValuemin,
    AriaValuenow,
    AriaValuetext,
    ARIAState(ARIAState),
}

pub enum ARIAPropertyDefinitionType {
    String,
    Id,
    IdList,
    Integer,
    Number,
    Boolean,
    Token,
    TokenList,
    Tristate,
}

pub struct ARIAPropertyDefinition {
    pub key: ARIAProperty,
    pub type_: ARIAPropertyDefinitionType,
    pub values: Option<&'static [&'static str]>,
    pub allow_undefined: Option<bool>,
}

#[cfg(test)]
impl Serialize for ARIAPropertyDefinition {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("ARIAProperty", 4)?;
        state.serialize_field(
            "key",
            match self.key {
                ARIAProperty::AriaActivedescendant => "aria-activedescendant",
                ARIAProperty::AriaAtomic => "aria-atomic",
                ARIAProperty::AriaAutocomplete => "aria-autocomplete",
                ARIAProperty::AriaBraillelabel => "aria-braillelabel",
                ARIAProperty::AriaBrailleroledescription => "aria-brailleroledescription",
                ARIAProperty::AriaColcount => "aria-colcount",
                ARIAProperty::AriaColindex => "aria-colindex",
                ARIAProperty::AriaColspan => "aria-colspan",
                ARIAProperty::AriaControls => "aria-controls",
                ARIAProperty::AriaCurrent => "aria-current",
                ARIAProperty::AriaDescribedby => "aria-describedby",
                ARIAProperty::AriaDescription => "aria-description",
                ARIAProperty::AriaDetails => "aria-details",
                ARIAProperty::AriaDropeffect => "aria-dropeffect",
                ARIAProperty::AriaErrormessage => "aria-errormessage",
                ARIAProperty::AriaFlowto => "aria-flowto",
                ARIAProperty::AriaHaspopup => "aria-haspopup",
                ARIAProperty::AriaKeyshortcuts => "aria-keyshortcuts",
                ARIAProperty::AriaLabel => "aria-label",
                ARIAProperty::AriaLabelledby => "aria-labelledby",
                ARIAProperty::AriaLevel => "aria-level",
                ARIAProperty::AriaLive => "aria-live",
                ARIAProperty::AriaModal => "aria-modal",
                ARIAProperty::AriaMultiline => "aria-multiline",
                ARIAProperty::AriaMultiselectable => "aria-multiselectable",
                ARIAProperty::AriaOrientation => "aria-orientation",
                ARIAProperty::AriaOwns => "aria-owns",
                ARIAProperty::AriaPlaceholder => "aria-placeholder",
                ARIAProperty::AriaPosinset => "aria-posinset",
                ARIAProperty::AriaReadonly => "aria-readonly",
                ARIAProperty::AriaRelevant => "aria-relevant",
                ARIAProperty::AriaRequired => "aria-required",
                ARIAProperty::AriaRoledescription => "aria-roledescription",
                ARIAProperty::AriaRowcount => "aria-rowcount",
                ARIAProperty::AriaRowindex => "aria-rowindex",
                ARIAProperty::AriaRowspan => "aria-rowspan",
                ARIAProperty::AriaSetsize => "aria-setsize",
                ARIAProperty::AriaSort => "aria-sort",
                ARIAProperty::AriaValuemax => "aria-valuemax",
                ARIAProperty::AriaValuemin => "aria-valuemin",
                ARIAProperty::AriaValuenow => "aria-valuenow",
                ARIAProperty::AriaValuetext => "aria-valuetext",
                ARIAProperty::ARIAState(ARIAState::AriaBusy) => "aria-busy",
                ARIAProperty::ARIAState(ARIAState::AriaChecked) => "aria-checked",
                ARIAProperty::ARIAState(ARIAState::AriaDisabled) => "aria-disabled",
                ARIAProperty::ARIAState(ARIAState::AriaExpanded) => "aria-expanded",
                ARIAProperty::ARIAState(ARIAState::AriaGrabbed) => "aria-grabbed",
                ARIAProperty::ARIAState(ARIAState::AriaHidden) => "aria-hidden",
                ARIAProperty::ARIAState(ARIAState::AriaInvalid) => "aria-invalid",
                ARIAProperty::ARIAState(ARIAState::AriaPressed) => "aria-pressed",
                ARIAProperty::ARIAState(ARIAState::AriaSelected) => "aria-selected",
            },
        )?;
        state.serialize_field(
            "type",
            match self.type_ {
                ARIAPropertyDefinitionType::String => "string",
                ARIAPropertyDefinitionType::Id => "id",
                ARIAPropertyDefinitionType::IdList => "idlist",
                ARIAPropertyDefinitionType::Integer => "integer",
                ARIAPropertyDefinitionType::Number => "number",
                ARIAPropertyDefinitionType::Boolean => "boolean",
                ARIAPropertyDefinitionType::Token => "token",
                ARIAPropertyDefinitionType::TokenList => "tokenlist",
                ARIAPropertyDefinitionType::Tristate => "tristate",
            },
        )?;
        match &self.values {
            Some(values) => {
                state.serialize_field("values", &values)?;
            }
            None => {
                state.skip_field("values")?;
            }
        };
        match self.allow_undefined {
            Some(allow_undefined) => {
                state.serialize_field("allow_undefined", &allow_undefined)?;
            }
            None => {
                state.skip_field("allow_undefined")?;
            }
        };

        state.end()
    }
}

#[cfg_attr(test, derive(Serialize))]
pub enum ARIAPropertyCurrent {
    Page,
    Step,
    Location,
    Date,
    Time,
    True,
    False,
}

// These constraints are drawn from the mapping between ARIA and HTML:
// https://www.w3.org/TR/html-aria
#[cfg_attr(test, derive(Serialize))]
pub struct ARIARoleRelationConceptAttributeConstraints {
    // The attribute does not exist on the node: <a>
    pub undefined: bool,
    // The attribute has a value: <a b="c">
    pub set: bool,
    pub gt1: bool,
}

#[cfg_attr(test, derive(Serialize))]
pub struct ARIARoleRelationConceptAttribute {
    pub name: String,
    pub value: Option<String>,
    pub constraints: Option<ARIARoleRelationConceptAttributeConstraints>,
}

// These constraints are drawn from the mapping between ARIA and HTML:
// https://www.w3.org/TR/html-aria
#[cfg_attr(test, derive(Serialize))]
pub struct ARIARoleRelationConceptConstraints {
    pub scoped_to_the_body_element: bool,
    pub scoped_to_the_main_element: bool,
    pub scoped_to_a_sectioning_root_element_other_than_body: bool,
    pub scoped_to_a_sectioning_content_element: bool,
    pub direct_descendant_of_document: bool,
    pub direct_descendant_of_ol: bool,
    pub direct_descendant_of_ul: bool,
    pub direct_descendant_of_menu: bool,
    pub direct_descendant_of_details_element_with_the_open_attribute_defined: bool,
    pub ancestor_table_element_has_table_role: bool,
    pub ancestor_table_element_has_grid_role: bool,
    pub ancestor_table_element_has_treegrid_role: bool,
    pub the_size_attribute_value_is_greater_than_1: bool,
    pub the_multiple_attribute_is_not_set_and_the_size_attribute_does_not_have_a_value_greater_than_1:
        bool,
    pub the_list_attribute_is_not_set: bool,
}

/* The concept in a related domain that informs behavior mappings.
** Related domains include: HTML, "Device Independence Delivery Unit", XForms,
** and ARIA to name a few.
**/
pub struct ARIARoleRelationConcept {
    pub name: &'static str,
    pub attributes: Option<Vec<ARIARoleRelationConceptAttribute>>,
    pub constraints: Option<ARIARoleRelationConceptConstraints>,
}

#[cfg(test)]
impl Serialize for ARIARoleRelationConcept {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("ARIARoleRelation", 4)?;

        match &self.attributes {
            Some(values) => {
                state.serialize_field("attributes", &values)?;
            }
            None => {
                state.skip_field("attributes")?;
            }
        };

        match &self.constraints {
            Some(values) => {
                state.serialize_field("constraints", &values)?;
            }
            None => {
                state.skip_field("constraints")?;
            }
        };

        state.end()
    }
}

#[cfg_attr(test, derive(Serialize))]
pub struct ARIARoleRelation {
    pub module: Option<&'static str>,
    pub concept: Option<ARIARoleRelationConcept>,
}

pub type ARIAProps = OrderedMap<&'static str, &'static Option<ARIAPropertyCurrent>>;

pub struct ARIARoleDefinition {
    /* Abstract roles may not be used in HTML. */
    pub is_abstract: bool,
    pub accessible_name_required: bool,
    /* The concepts in related domains that inform behavior mappings. */
    pub base_concepts: Option<&'static [&'static ARIARoleRelation]>,
    /* Child presentational roles strip child nodes of roles and flatten the content to text. */
    pub children_presentational: bool,
    pub name_from: Option<&'static [&'static ARIANameFromSources]>,
    /* aria-* properties and states disallowed on this role. */
    pub prohibited_props: Option<&'static [&'static ARIAProperty]>,
    /* aria-* properties and states allowed on this role. */
    pub props: Option<&'static ARIAProps>,
    /* The concepts in related domains that inform behavior mappings. */
    pub related_concepts: Option<&'static [ARIARoleRelation]>,
    pub require_context_role: Option<&'static [&'static ARIARole]>,
    pub required_context_role: Option<&'static [&'static ARIARole]>,
    pub required_owned_elements: Option<&'static [&'static [&'static ARIARole]]>,
    /* aria-* properties and states required on this role. */
    pub required_props: Option<&'static ARIAProps>,
    /* An array or super class "stacks." Each stack contains a LIFO list of
     ** strings correspond to a super class in the inheritance chain of this
     ** role. Roles may have more than one inheritance chain, which is why
     ** this property is an array of arrays and not a single array. */
    pub super_class: Option<&'static [&'static [&'static ARIARoleDefinitionKey]]>,
}

#[cfg(test)]
impl Serialize for ARIARoleDefinition {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("ARIARoleDefinition", 4)?;

        match &self.base_concepts {
            Some(values) => {
                state.serialize_field("base_concepts", &values)?;
            }
            None => {
                state.skip_field("base_concepts")?;
            }
        };

        match &self.name_from {
            Some(values) => {
                state.serialize_field("name_from", &values)?;
            }
            None => {
                state.skip_field("name_from")?;
            }
        };

        match &self.prohibited_props {
            Some(values) => {
                state.serialize_field("prohibited_props", &values)?;
            }
            None => {
                state.skip_field("prohibited_props")?;
            }
        };

        match &self.related_concepts {
            Some(values) => {
                state.serialize_field("related_concepts", &values)?;
            }
            None => {
                state.skip_field("related_concepts")?;
            }
        };

        state.end()
    }
}

pub type RoleDefinitions = OrderedMap<&'static str, &'static ARIARoleDefinition>;
