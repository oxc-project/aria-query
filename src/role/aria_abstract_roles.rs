use crate::definition::{
    ARIAAbstractRole, ARIADocumentStructureRole, ARIANameFromSources, ARIARole, ARIARoleDefinition,
    ARIARoleDefinitionKey, ARIARoleRelation, ARIARoleRelationConcept, RoleDefinitions,
};
use phf::phf_ordered_map;

pub static ARIA_ABSTRACT_ROLES: RoleDefinitions = phf_ordered_map! {
    "command" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: None,
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Widget)]]),
    },
    "composite" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-activedescendant" => &None,
            "aria-disabled" => &None,
        }),
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Widget)]])
    },
    "input" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-disabled" => &None,
        }),
        related_concepts: Some(&[
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "input",
                    attributes: None,
                    constraints: None
                }),
                module: Some("XForms"),
            }
        ]),
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Widget)]])
    },
    "landmark" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: None,
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Structure), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Section)]])
    },
    "range" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-valuemax" => &None,
            "aria-valuemin" => &None,
            "aria-valuenow" => &None,
        }),
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Structure)]])
    },
    "roletype" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: None,
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-atomic" => &None,
            "aria-busy" => &None,
            "aria-controls" => &None,
            "aria-describedby" => &None,
            "aria-details" => &None,
            "aria-dropeffect" => &None,
            "aria-flowto" => &None,
            "aria-grabbed" => &None,
            "aria-hidden" => &None,
            "aria-keyshortcuts" => &None,
            "aria-label" => &None,
            "aria-labelledby" => &None,
            "aria-live" => &None,
            "aria-owns" => &None,
            "aria-relevant" => &None,
            "aria-roledescription" => &None,
        }),
        related_concepts: Some(&[
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "role",
                    attributes: None,
                    constraints: None
                }),
                module: Some("XHTML"),
            },
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "type",
                    attributes: None,
                    constraints: None
                }),
                module: Some("Dublin Core"),
            }
        ]),
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: None
    },
    "section" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: None,
        prohibited_props: None,
        props: None,
        related_concepts: Some(&[
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "frontmatter",
                    attributes: None,
                    constraints: None
                }),
                module: Some("DTB"),
            },
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "level",
                    attributes: None,
                    constraints: None
                }),
                module: Some("DTB"),
            },
            ARIARoleRelation {
                concept: Some(ARIARoleRelationConcept {
                    name: "level",
                    attributes: None,
                    constraints: None
                }),
                module: Some("SMIL"),
            },
        ]),
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Structure)]])
    },
    "sectionhead" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author, &ARIANameFromSources::Contents]),
        prohibited_props: None,
        props: None,
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Structure)]])
    },
    "select" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-orientation" => &None,
        }),
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Widget), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Composite)],
        &[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Structure), &ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::Section), &ARIARoleDefinitionKey::ARIARole(ARIARole::ARIADocumentStructureRole(ARIADocumentStructureRole::Group))]])
    },
    "structure" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: None,
        prohibited_props: None,
        props: None,
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType)]])
    },
    "widget" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: None,
        prohibited_props: None,
        props: None,
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType)]])
    },
    "window" => &ARIARoleDefinition {
        is_abstract: true,
        accessible_name_required: false,
        base_concepts: None,
        children_presentational: false,
        name_from: Some(&[&ARIANameFromSources::Author]),
        prohibited_props: None,
        props: Some(&phf_ordered_map! {
            "aria-modal" => &None,
        }),
        related_concepts: None,
        require_context_role: None,
        required_context_role: None,
        required_owned_elements: None,
        required_props: None,
        super_class: Some(&[&[&ARIARoleDefinitionKey::ARIAAbstractRole(ARIAAbstractRole::RoleType)]])
    },
};
