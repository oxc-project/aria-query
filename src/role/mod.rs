mod aria_abstract_roles;
use crate::definition::ARIARoleDefinition;
use std::collections::HashMap;

pub fn entries() -> HashMap<&'static str, &'static ARIARoleDefinition> {
    aria_abstract_roles::ARIA_ABSTRACT_ROLES
        .entries()
        .map(|(k, v)| (*k, *v))
        .collect()
}

pub fn for_each(mut callback: impl FnMut(&'static str, &'static ARIARoleDefinition)) {
    aria_abstract_roles::ARIA_ABSTRACT_ROLES
        .into_iter()
        .for_each(|(k, v)| callback(k, v));
}

pub fn get(name: &str) -> Option<&'static ARIARoleDefinition> {
    match aria_abstract_roles::ARIA_ABSTRACT_ROLES.get(name) {
        Some(v) => Some(v),
        None => None,
    }
}

pub fn has(name: &str) -> bool {
    aria_abstract_roles::ARIA_ABSTRACT_ROLES.contains_key(name)
}

pub fn keys() -> impl Iterator<Item = &'static str> {
    aria_abstract_roles::ARIA_ABSTRACT_ROLES.keys().copied()
}

pub fn values() -> impl Iterator<Item = &'static ARIARoleDefinition> {
    aria_abstract_roles::ARIA_ABSTRACT_ROLES.values().copied()
}

#[cfg(test)]
mod test {
    use crate::definition::ARIARoleDefinition;
    use crate::role;
    use insta::{assert_json_snapshot, Settings};

    #[test]
    fn snapshot_for_entries() {
        let roles_entries = role::entries();

        let mut settings = Settings::clone_current();
        settings.set_sort_maps(true);
        settings.bind(|| {
            assert_json_snapshot!(roles_entries);
        });
    }

    #[test]
    fn test_for_each() {
        let mut el_count = 0;
        role::for_each(|_, _| el_count += 1);
        assert_eq!(el_count, 12);
        let mut elements_list = Vec::new();
        role::for_each(|k, _| {
            elements_list.push(k);
        });
        assert_eq!(elements_list, role::keys().collect::<Vec<_>>());
    }

    #[test]
    fn test_get() {
        assert!(role::get("command").is_some());
        assert!(role::get("unknown").is_none());
    }

    #[test]
    fn test_has() {
        assert!(role::has("command"));
        assert!(!role::has("unknown"));
    }

    #[test]
    fn test_keys() {
        let keys = role::keys().collect::<Vec<_>>();
        for key in keys {
            assert!(role::entries().contains_key(key));
        }
    }

    #[test]
    fn snapshot_values() {
        let roles_values = role::values();
        let roles_values = roles_values.collect::<Vec<&'static ARIARoleDefinition>>();
        let mut settings = Settings::clone_current();
        settings.set_sort_maps(true);
        settings.bind(|| {
            assert_json_snapshot!(roles_values);
        });
    }
}
