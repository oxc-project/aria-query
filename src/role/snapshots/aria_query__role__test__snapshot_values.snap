---
source: src/role/mod.rs
expression: roles_values
---
[
  {
    "name_from": [
      "Author"
    ]
  },
  {
    "name_from": [
      "Author"
    ]
  },
  {
    "name_from": [
      "Author"
    ],
    "related_concepts": [
      {
        "module": "XForms",
        "concept": {}
      }
    ]
  },
  {
    "name_from": [
      "Author"
    ]
  },
  {
    "name_from": [
      "Author"
    ]
  },
  {
    "related_concepts": [
      {
        "module": "XHTML",
        "concept": {}
      },
      {
        "module": "Dublin Core",
        "concept": {}
      }
    ]
  },
  {
    "related_concepts": [
      {
        "module": "DTB",
        "concept": {}
      },
      {
        "module": "DTB",
        "concept": {}
      },
      {
        "module": "SMIL",
        "concept": {}
      }
    ]
  },
  {
    "name_from": [
      "Author",
      "Contents"
    ]
  },
  {
    "name_from": [
      "Author"
    ]
  },
  {},
  {},
  {
    "name_from": [
      "Author"
    ]
  }
]
